state/executor.go

```diff
@@ -133,11 +212,19 @@ func (e *Executor) ProcessBlock(parentRoot types.Hash, block *types.Block, block
 	txn.block = block
 	for _, t := range block.Transactions {
 		if err := txn.Write(t); err != nil {
+			// Because block processing termination is here,
+			// the executor needs to handle any "stale" pending events,
+			// since every transaction in this block is discarded
+			cleanDiscardedEvents(block.Transactions, block.Number())
+
 			return nil, err
 		}
 	}
 	_, root := txn.Commit()
 
+	// Checks are passed, do staking logic if possible
+	commitApprovedEvents(block.Transactions, block.Number())
```

Comment:
If `commitApprovedEvents` are after `txn.Commit()` then I'm afraid this attack might be possible:
1. I send n unstaking transactions to be mined in a single block
2. I make sure that the sum of my transactions isn't greater than the staked amount of the whole network
3. I receive n * myStakedBalance to my account.

In the end `commitApprovedEvents` will just have n calls to `ResetStake` which will pass.

Reply:
If this turns out to be true, it would be amazing to write a test that covers this before fixing it.

Reply:
@zivkovicmilos GJ on the test, exactly what I was hoping for 💯